packages/alfa-rules/src/sia-r15/rule.ts

```diff
@@ -107,3 +106,71 @@ export namespace Outcomes {
     )
   );
 }
+
+type Source = "srcdoc" | "src" | "invalid" | "nothing";
+
+class EmbeddedResource implements Equatable, Hashable {
+  public static of(source: Source, value?: string): EmbeddedResource {
+    return new EmbeddedResource(source, Option.from(value));
+  }
+
+  private readonly _source: Source;
+  private readonly _value: Option<string>;
+
+  private constructor(source: Source, value: Option<string>) {
+    this._source = source;
+    this._value = value;
+  }
+
+  public get value(): Option<string> {
+    return this._value;
+  }
+
+  public equals(value: unknown): value is this {
+    return (
+      value instanceof EmbeddedResource &&
+      this._source === value._source &&
+      this._value.equals(value._value)
+    );
+  }
+
+  public hash(hash: Hash) {
+    Hash.writeString(hash, this._source);
+    this._value.hash(hash);
+  }
+
+  public toString() {
+    return `{source: ${this._source}, value: ${this._value}}`;
+  }
+}
+
+/**
+ * @see https://html.spec.whatwg.org/multipage/iframe-embed-object.html#process-the-iframe-attributes
+ */
+function embeddedResource(iframe: Element): EmbeddedResource {
+  // srcdoc takes precedence.
+  if (iframe.attribute("srcdoc").isSome()) {
+    return EmbeddedResource.of(
+      "srcdoc",
+      iframe.attribute("srcdoc").get().value
+    );
+  }
+
+  // Otherwise, grab the src attribute.
+  function getUrl(value: string): EmbeddedResource {
+    try {
+      return EmbeddedResource.of("src", new URL(value).href);
```

Comment:
We can resolve the reference against the URL of the `Response` instance.

Reply:
Would be cool, but I don't know where to access that 🙃

Reply:
See https://github.com/Siteimprove/alfa/issues/420#issuecomment-701479693.